clients/rust/client/src/api.rs

```diff
@@ -0,0 +1,271 @@
+            .map_err(|err| GammaError::FailedSendingRequest {
+                context: context.into(),
+                err,
+            })?;
+
+        let status = response.status();
+        if !status.is_success() {
+            if status == 404 {
+                return Err(GammaError::NotFoundResponse {
+                    context: context.into(),
+                });
+            }
+
+            let body_str = response
+                .text()
+                .await
+                .unwrap_or("Failed to read response body".into());
+
+            return Err(GammaError::ErrorResponse {
+                context: context.into(),
+                status,
+                body_str,
+            });
+        }
+
+        let body: T =
+            response
+                .json()
+                .await
+                .map_err(|err| GammaError::FailedToDeserializeResponse {
+                    context: context.into(),
+                    error: err,
+                })?;
+
+        Ok(body)
+    }
+}
+
+/// A group in gamma (e.g. digIT'21).
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GammaGroup {
+    /// A unique identifier for the group.
+    pub id: Uuid,
+    /// The name of the group.
+    pub name: String,
+    /// The pretty name of the group.
+    pub pretty_name: String,
+    /// The supergroup this group belongs to (e.g. digIT).
+    pub super_group: GammaSuperGroup,
+}
+
+/// A supergroup in gamma (e.g. digIT).
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GammaSuperGroup {
+    /// A unique identifier for the supergroup.
+    pub id: Uuid,
+    /// The name of the supergroup (e.g. "digit").
+    pub name: String,
+    /// The pretty name of the supergroup (e.g. "digIT").
+    pub pretty_name: String,
+    /// The type of supergroup this is.
+    #[serde(rename = "type")]
+    pub group_type: GammaSuperGroupType,
+    /// The swedish description of the supergroup.
+    pub sv_description: String,
+    /// The english description of the supergroup.
+    pub en_description: String,
+}
+
+/// A type of supergroup.
+/// Note: In gamma these are generic strings and can be created and deleted through the GUI,
+/// for this reason an `Other` option is provided as a cath all and the remaining options
+/// are simply the ones currently in use at the time of writing.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub enum GammaSuperGroupType {
+    /// An alumni group (not active).
+    Alumni,
+    /// A committee within the IT divison.
+    Committee,
+    /// A society recognized by the IT division.
+    Society,
+    /// Functionary groups within the IT division (e.g. auditors).
+    Functionaries,
+    /// Gamma administrators.
+    Admin,
+    /// A type that of group that is not specificly supported by this client.
+    #[serde(untagged)]
+    Other(String),
+}
```

Comment:
I do not agree that `cthit/Gamma` should provide clients that have super group types specified. The long-term goal of Gamma is to be an account service for student divisions other than just IT, and the clients it provides should support that goal. 

I'm aware that this will affect how typing works with Rust, but I still do not think that should be prioritized over the long-term goal of Gamma.

Reply:
I can partially agree with this as super group types are defined dynamically in Gamma 2.0 so technically all of the specified groups could be missing and only Other would exist. What I don't know however is how we can do this better, for our use it is very handy to have an enum that represents the common types that we actually use.
If we just remove the comments specifying the IT division this would be applicable to basically all other student divisions on Chalmers at least? As for the Admin type that should probably be removed completely as it is not used ether way.

Reply:
Let's talk further on a future workshop how to handle clients in general, since I think a rust client will not be the last implemented client.